ayon_server/operations/project_level/__init__.py

```diff
@@ -133,17 +130,26 @@ async def _process_operations(
         )
 
         try:
-            entity, evt, response = await _process_operation(
-                project_name,
-                user,
-                operation,
-                transaction=transaction,
-            )
-            if evt is not None:
-                events.extend(evt)
-            result.append(response)
-            if entity.entity_type not in [e.entity_type for e in to_commit]:
-                to_commit.append(entity)
+            # This is a neat trick. transaction() will try
+            # to reuse the current transaction if it exists,
+            # but create a new one if it doesn't.
+
+            # This way, every operation is ensured to run in a transaction,
+            # but main process may wrap everything in a single one
+            # to commit all operations at once.
```

Comment:
I'm not sure what the size of the of the 'transaction' may become at this point, but if it may grow large with more and more stacking into a transaction. Then it may be good to be aware of potential downsides due to it, e.g. [locking](https://dba.stackexchange.com/a/273905). Rows touched by the transaction will remain locked for the remainder duration of the transaction. This may make sense, but beware that (according to that page?) even reading the row may be blocked _during_ that transaction?

In short, transactions may limit the possibility of concurrency.

_I know nothing of databases NOR of this part of the code, so I'm likely saying very stupid stuff here._